core/Pos/Core/Vss.hs

```diff
@@ -71,22 +86,75 @@ checkCertSign VssCertificate {..} =
 getCertId :: VssCertificate -> StakeholderId
 getCertId = addressHash . vcSigningKey
 
--- | Safe constructor of 'VssCertificatesMap'. TODO: wrap into newtype.
-mkVssCertificatesMap :: [VssCertificate] -> VssCertificatesMap
-mkVssCertificatesMap = HM.fromList . map toCertPair
+-- | Safe constructor of 'VssCertificatesMap'. It doesn't allow certificates
+-- with duplicate signing keys or with duplicate 'vcVssKey's.
+mkVssCertificatesMap
+    :: MonadFail m
+    => [VssCertificate] -> m VssCertificatesMap
+mkVssCertificatesMap certs = do
+    unless (allDistinct (map vcSigningKey certs)) $
+        fail "mkVssCertificatesMap: two certs have the same signing key"
+    unless (allDistinct (map vcVssKey certs)) $
+        fail "mkVssCertificatesMap: two certs have the same VSS key"
+    pure $ UnsafeVssCertificatesMap (HM.fromList (map toCertPair certs))
+  where
+    toCertPair vc = (getCertId vc, vc)
+
+-- | A convenient constructor of 'VssCertificatesMap' that throws away
+-- certificates with duplicate signing keys or with duplicate 'vcVssKey's.
+mkVssCertificatesMapLossy :: [VssCertificate] -> VssCertificatesMap
+mkVssCertificatesMapLossy =
+    UnsafeVssCertificatesMap . HM.fromList .
+    map toCertPair . nubOrdOn vcVssKey
   where
     toCertPair vc = (getCertId vc, vc)
 
+-- | A map with a single certificate is always valid so this function is
+-- safe to use in case you have one certificate and want to create a map
+-- from it.
+mkVssCertificatesMapSingleton :: VssCertificate -> VssCertificatesMap
+mkVssCertificatesMapSingleton vc =
+    UnsafeVssCertificatesMap $ HM.singleton (getCertId vc) vc
```

Comment:
`mkVssCertificatesMapLossy [vc]` or `mkVssCertificatesMapLossy . one` might be a bit better, but it's subjective and minor.